hudi-common/src/main/java/org/apache/hudi/common/model/HoodieFileGroup.java

```diff
@@ -178,7 +203,15 @@ public Option<HoodieBaseFile> getLatestDataFile() {
    * Obtain the latest file slice, upto a instantTime i.e <= maxInstantTime.
    */
   public Option<FileSlice> getLatestFileSliceBeforeOrOn(String maxInstantTime) {
-    return Option.fromJavaOptional(getAllFileSlices().filter(slice -> compareTimestamps(slice.getBaseInstantTime(), LESSER_THAN_OR_EQUALS, maxInstantTime)).findFirst());
+    return getLatestFileSliceBeforeOrOn(maxInstantTime, false);
+  }
+
+  /**
+   * Obtain the latest file slice, upto a instantTime i.e <= maxInstantTime.
+   */
+  public Option<FileSlice> getLatestFileSliceBeforeOrOn(String maxInstantTime, boolean includePending) {
```

Comment:
We should be cautious to expose inflight file slices on the filesystem view, the MVCC completeness are somehow broken for all the readers.

Reply:
In general, we should add core API only for tesing purpose

Reply:
I agree. Core writer flow does not use them, some are used by query engines. They are able to get it working in a weird way, like some times they pass active timeline to the file system view and sometimes they pass write timeline etc.

Reply:
Yeah, the timeline of the filesystem view is a chaos, can we avoid introducing these testing APIs on the filesystem view and just refactors the tests as correct?